### PR TITLE
[rtsan][compiler-rt] Prevent UB hang in rtsan lock unit tests

### PR DESCRIPTION
**Pull request by cjappl**
It is undefined behavior to lock or unlock an uninitialized lock, and unlock a lock which isn't locked.

Later in one of my dev branches this was hanging on ubuntu.

This review introduces a fixture to set up and tear down the locks where appropriate, and separates them into two tests (realtime death and non realtime survival) so each test is guaranteed a fresh lock.

### CODE REVIEW

File: compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp [hunk 1]
```diff
@@ -328,26 +328,64 @@ TEST(TestRtsanInterceptors, PthreadCreateDiesWhenRealtime) {
...
+  void TearDown() override {
+    if (is_locked) {
+      Unlock();
+    }
+    pthread_mutex_destroy(&mutex);
+  }
+
+  void Lock() {
+    ASSERT_TRUE(!is_locked);
     pthread_mutex_lock(&mutex);
```
Comment (cjappl) [inline compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp:347]: Notice here, for instance, we were locking an uninitialized lock

File: compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp [hunk 2]
```diff
@@ -328,26 +328,64 @@ TEST(TestRtsanInterceptors, PthreadCreateDiesWhenRealtime) {
...
+TEST_F(PthreadMutexLockTest, PthreadMutexLockSurvivesWhenNotRealtime) {
+  auto Func = [this]() { Lock(); };
+
   ExpectNonRealtimeSurvival(Func);
 }
 
-TEST(TestRtsanInterceptors, PthreadMutexUnlockDiesWhenRealtime) {
-  auto Func = []() {
-    pthread_mutex_t mutex{};
-    pthread_mutex_unlock(&mutex);
```
Comment (cjappl) [inline compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp:344]: Unlocking an uninited, and not-locked lock

**Comment by cjappl** [comment]
CC @davidtrevelyan

File: compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp [hunk 3]
```diff
@@ -328,26 +328,64 @@ TEST(TestRtsanInterceptors, PthreadCreateDiesWhenRealtime) {
...
-    pthread_mutex_t mutex{};
+class PthreadMutexLockTest : public ::testing::Test {
+protected:
+  void SetUp() override {
+    pthread_mutex_init(&mutex, nullptr);
+    is_locked = false;
+  }
+
+  void TearDown() override {
+    if (is_locked) {
```
Comment (MaskRay) [inline compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp:339]: delete braces per https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements

**Review by MaskRay — APPROVED** [review]
(I'm starting a 3-week vacation this Friday and will have limited availability.)

File: compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp (hunk 3 above)
Comment (cjappl) [reply compiler-rt/lib/rtsan/tests/rtsan_test_interceptors.cpp]: One of these reviews I'll catch all of them before submitting...

Enjoy your vacation! Thanks again for all your help with RTSan, checking out all our reviews. I very much appreciate you taking the time.

Talk to you when you get back

**Pull request merged**